samples: Bluetooth: Add Direction Finding CTE Tx sample application

ppryga-nordic opened the pull request:
The PR includes:
- fix for not switching antennas in AoD mode,
- initialization of GPIOs responsible for driving antenna switches
- initialization of default antenna used for PDU transmission/reception
- sample application that transmits CTE with periodic advertising PDUs.

ppryga-nordic committed:
samples: Bluetooth: Add DF connectionless Beacon sample application

Add an application that uses Direction Finding API for
transmission of CTE in connectionless mode (periodic adverising PDUs).

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: Kconfig: Disable ant switching when CTE Rx not supp

Anternna switching should be disabled when CTE Rx is not supported
or disabled. Added missing dependency on BT_CTLR_DF_CTE_RX.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: LLL: Fix not switching antennas in AoD mode

Antenna switching done by Radio with use of GPIOs provided in
PSEL.DFEGPIO registers. Registers were initialized once during
controller initialization.
At preparation of new radio event, Radio peripheral is powered
down and up to resset its registers. It is related with multi
protocol handling and possible leftovers in Radio registers.

Due to that, PSEL.DFEGPIO registers should be initialized
at the preparation of every event that will transmit
or receive CTE and run antenna switching.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga-nordic committed:
Bluetooth: controller: LLL: DF add initialization of ant switching GPIOs

Add initialization of GPIOs that are used for antenna switching.
GPIOs provided to DFE extension in Radio peripheral, to drive
antenna switching, should be initialized in GPIO peripheral beforehand.
The initialization is optional and may be disabled by Kconfig.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Gentlemen can you do a review of the PR?

jhedberg commented on samples/bluetooth/df_per_adv_beacon/src/main.c:
```diff
@@ -0,0 +1,208 @@
+
+static void bt_ready(int err)
+{
+	printk("Bluetooth initialization...\n");
```
I guess it'd be more appropriate to have this `printk` in `main()` right before you call `bt_enable()`?

jhedberg replied:
also, I suppose you could drop the `\n` if you want

cvinayak replied:
Also, use the main thread, do not use this callback function which could be a different thread (system work queue?).

ppryga replied:
Changed according to suggestion.

jhedberg commented on samples/bluetooth/df_per_adv_beacon/README.rst:
```diff
@@ -0,0 +1,65 @@
+Overview
+********
+
+A simple application demonstrating the BLE Direction Finding CTE Broadcaster functionality by sending Constant Tone Extension with periodic advertising PDUs.
```
Not 100% sure since I'm reviewing this in a browser, but doesn't this go beyond the 100 character line length limit (same for other parts of this file as well)

ppryga replied:
It was over 100 characters in a line. I've thought that RST files are not limited to 100 chars per line. 

Line breaks added to fit into the limit.

joerchan commented on samples/bluetooth/df_per_adv_beacon/src/main.c:
```diff
@@ -0,0 +1,208 @@
+{
+	int err;
+
+	printk("Periodic advertising enable...");
```
Minor: I think I would prefer to have printing done consistenly with the other samples, like this:
```
	err = bt_le_adv_start(BT_LE_ADV_CONN_NAME, ad, ARRAY_SIZE(ad), NULL, 0);
	if (err) {
		printk("Advertising failed to start (err %d)\n", err);
		return;
	}

	printk("Advertising successfully started\n");

ppryga replied:
I've followed @cvinayak sample to be consistent: https://github.com/zephyrproject-rtos/zephyr/blob/3fe33fe3883ffabb8a92bc59debd605e50c26ad7/samples/bluetooth/periodic_sync/src/main.c#L203-L209

cvinayak replied:
The `printk` before the API call help debugging if an API hangs (which does often when developing the controller ;-) )

joerchan replied:
Haha, Isn't enabling logging better suited for that purpose?

As I said this is minor, I would have preferred to follow the same style on all of them, but make your own decision.

ppryga replied:
I'll follow BT samples style,

joerchan commented on samples/bluetooth/df_per_adv_beacon/src/main.c:
```diff
@@ -0,0 +1,208 @@
+	bt_addr_le_t addr = {0};
+	size_t count = 1;
+
+	bt_id_get(&addr, &count);
```
This would be incorrect if BT_PRIVACY was enabled.
Better to use `bt_le_ext_adv_oob_get_local` to retrieve the advertiser address.

ppryga replied:
Thank you for the hint. Changed to use bt_le_ext_adv_oob_get_local.

joerchan commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -262,9 +274,20 @@ void radio_df_ant_switch_spacing_set_4us(void)
 {
-	NRF_RADIO->SWITCHPATTERN = pattern;
+	/* DFE extension in radio uses:
+	 * - SWITCHPATTER[0] for idle period (PDU Tx/Rx),
```
Missing `N`? `SWITCHPATTERN`

ppryga replied:
This is correct. The SWITCHPATTER is kind of a pointer to underlying buffer. Every write moves the pointer to next free memory location. When read returns number of written elements. If want to start writing from beginning you should write to CLEARPATTERN register to before.

joerchan commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -262,9 +274,20 @@ void radio_df_ant_switch_spacing_set_4us(void)
+	 * DTS property dfe_pdu_ant. This limits number of supported antenna
+	 * switch patterns by one.
+	 */
+	NRF_RADIO->SWITCHPATTERN = DT_PROP(RADIO, dfe_pdu_ant);
```
I'm not able to match this with the comment above, shouldn`t SWITCHPATTERN be indexed?

ppryga replied:
This is correct. The SWITCHPATTER is kind of a pointer to underlying buffer. Every write moves the pointer to next free memory location. When read returns number of written elements. If want to start writing from beginning you should write to CLEARPATTERN register to before.

joerchan replied:
Ok, thank you.

cvinayak commented on samples/bluetooth/df_per_adv_beacon/src/main.c:
```diff
@@ -0,0 +1,208 @@
+	}
+	printk("success\n");
+
+	adv_set = advertising_set_create();
+	if (!adv_set) {
+		return;
+	}
+
+	err = cte_params_set(adv_set, &cte_params);
+	if (err) {
+		return;
+	}
+
+	err = periodic_advertising_params_set(adv_set);
+	if (err) {
+		return;
+	}
+
+	err = cte_enable(adv_set);
+	if (err) {
+		return;
+	}
+
+	err = periodic_advertising_enable(adv_set);
+	if (err) {
+		return;
+	}
```
No need of second level of abstraction in the sample application, users should see the direct use of Bluetooth Host API use. (You are conveying that the API signatures are not good enough, needing abstractions).

ppryga replied:
It is a kind of leftover. I used the code in a more complicated scenarios, so I wanted to avoid multiple printk() with debug information. I'll remove 2nd level functions.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -7,6 +7,7 @@
 #include <stdint.h>
 #include <errno.h>
 #include <devicetree.h>
+#include <drivers/gpio.h>
```
How much of CPU use does GPIO API's use in ISR add to Radio IRQ? Remember, Radio ISR (and LLL context) is to run in Zero-Latency ISRs. Please consider use of nRFx functions unless the caller context is Thread.

ppryga replied:
> How much of CPU use does GPIO API's use in ISR add to Radio IRQ? Remember, Radio ISR (and LLL context) is to run in Zero-Latency ISRs. Please consider use of nRFx functions unless the caller context is Thread.

@cvinayak 
GPIO peripheral is used only by the function: `radio_df_ant_switching_gpios_cfg()` to configure pins selected for antenna switching. This function is called once in `lll_df_init()`. There is no direct use of GPIO peripheral by the code in ISR.

I'll change use of Zephyr GPIO API to nrfx api.

cvinayak reviewed (changes requested):
call the sample `direction_beacon`, users can understand the sample better at first look.

ppryga commented on samples/bluetooth/df_per_adv_beacon/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+)
+
+if (CONFIG_BT_CTLR_DF_ANT_SWITCH_TX)
+set(DTC_OVERLAY_FILE "${CMAKE_CURRENT_SOURCE_DIR}/boards/nrf52833dk_nrf52833.overlay")
```
I've added it by use of condition to show that information provided by nrf52833dk_nrf52833.overlay are required only for Direction Finding in AoD mode (when antennas are switched). For simple beacons working in AoA mode the overlay is not needed.

ppryga replied:
@tejlmand have you got an opinion about recommended practice in this regard?

mbolivar-nordic replied:
Unless I am missing something, this is not correct. Adding to DTC_OVERLAY_FILE at this point in the build will have no effect. You need to set DTC_OVERLAY_FILE before `find_package(Zephyr)`, since DTC_OVERLAY_FILE is processed by boilerplate.cmake during the `find_package()` and is ignored afterwards.

I understand this is a chicken-and-egg problem since you want to handle this after Kconfig, which runs after dtc.

I see no way to resolve this within our current build system constraints except for documentation.

tejlmand replied:
> @tejlmand have you got an opinion about recommended practice in this regard?

No, this is not a correct way.
What @mbolivar-nordic  describes is fully correct, but also remember that placing an overlay in `<sample>/boards` folder means it will automatically be sourced when building for that board, regardless of the Kconfig setting.
 
So when exactly is `nrf52833dk_nrf52833.overlay` needed ?
1. Always when building for the `nrf52833dk_nrf52833` board ?
2. Always when `CONFIG_BT_CTLR_DF_ANT_SWITCH_TX=y` regardless of the board in use ?
3. Only when building for the `nrf52833dk_nrf52833` board and `CONFIG_BT_CTLR_DF_ANT_SWITCH_TX=y` ?

mbolivar-nordic replied:
> You need to set DTC_OVERLAY_FILE before `find_package(Zephyr)`, since DTC_OVERLAY_FILE is processed by boilerplate.cmake during the `find_package()` and is ignored afterwards.

FYI, I've sent a pull request which will detect and warn on invalid settings like this: https://github.com/zephyrproject-rtos/zephyr/pull/32556

mbolivar-nordic replied:
> but also remember that placing an overlay in `<sample>/boards` folder means it will automatically be sourced when building for that board, regardless of the Kconfig setting.

Good point -- I missed that. So it has no effect, but it's redundant :)

ppryga replied:
@mbolivar-nordic @tejlmand thank you very much for your input!

> So when exactly is nrf52833dk_nrf52833.overlay needed ?
At the moment it is needed for only when building for  `nrf52833dk_nrf52833` board and CONFIG_BT_CTLR_DF_ANT_SWITCH_TX=y
But it will change in future, when DF support is added to other boards. Then the overlay might be shared between all those boards.

tejlmand replied:
> At the moment it is needed for only when building for nrf52833dk_nrf52833 board and CONFIG_BT_CTLR_DF_ANT_SWITCH_TX=y
But it will change in future, when DF support is added to other boards. Then the overlay might be shared between all those boards.

in that case, we could also do it this way:
```
west build -b nrf52833dk_nrf52833 samples/bluetooth/df_per_adv_beacon -- -DCONF_FILE=prj_aod.conf
```
and then place the overlay specific for the build type: `boards/ nrf52833dk_nrf52833_aod.overlay`

But that would swap `aoa` <-> `aod` as to which is the default configuration.
And require this pr also: https://github.com/zephyrproject-rtos/zephyr/pull/32436

See more about this here: https://docs.zephyrproject.org/latest/guides/kconfig/setting.html#the-initial-configuration

ppryga replied:
@tejlmand But I can't wait until the PR: https://github.com/zephyrproject-rtos/zephyr/pull/32436 is merged... My PR must go not later than tomorrow. I need it to prepare another PR in ncs/zephyr that must go by the Friday.

ppryga replied:
@tejlmand This is last piece that is missing here. What do you think about leaving the `nrf52833dk_nrf52833_aod.overlay` to be included into the build independently from `CONFIG_BT_CTLR_DF_ANT_SWITCH_TX=y` and update the sample when PR: https://github.com/zephyrproject-rtos/zephyr/pull/32436 is ready?

ppryga-nordic committed:
Bluetooth: host: direction: Remove compilation warning

Remove compilation warning caused by currently not used
hci_df_set_conn_cte_tx_param function.
Code is sorrounded by #ifdef CONFIG_BT_CTLR_DF_CONN_CTE_RSP.
The CONFIG_BT_CTLR_DF_CONN_CTE_RSP is changed to be disabled
by default (until complete implementation of the feature is
provided).

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented on samples/bluetooth/df_per_adv_beacon/CMakeLists.txt:
```diff
@@ -0,0 +1,39 @@
+
+cmake_minimum_required(VERSION 3.13.1)
+
+if (NOT DF_MODE)
+  message("DF_MODE is not being defined on the CMake command-line in the environment or by the app\n"
+          "Setting DF_MODE to default value: AOD")
+  set(DF_MODE "AOD")
+endif()
```
It seems to me that it can't be done that way because there is a de-selected entry: https://github.com/zephyrproject-rtos/zephyr/blob/3dd35c202be239884a913d78780eda45ca3c810f/samples/bluetooth/df_per_adv_beacon/df_aoa.conf#L6

Although I can add Kconfig file with two entries `DF_AOA_MODE`, `DF_AOD_MODE` and according to selected one include appropriate *.conf file. 
I'd like to leave separate conf files because it provides complete Kconfig required to use:
- AoA mode
- AoA and AoD modes

joerchan replied:
Yeah, everything is `default y`, maybe they shouldn't have been?

You can also change the default locally in the Kconfig, and add default n.

I think the best would have been that the features didn't default to 'y' and you only turned them on when you actually needed to, if the prj.conf have to turn off all these features it's not a good default selection IMO.

Doing prj.conf manipulation is not a good idea as far as I know, but maybe @tejlmand has a different opinion on this?

ppryga replied:
Supported DF features are set by default to y. Use of e.g. only AoA mode is restricting functionality e.g. to decrease code size. So if one enables DF all supported features are available. From sample application point of view this is most desired configuration. I've provided AoA just to show that there is an option to do so.

ppryga replied:
@tejlmand have you got a suggestion about it?

tejlmand replied:
I see you write following in the docs:
> There are two build configuration provided with this sample:
> - df_aoa.config - enables Angle of Arrival mode
> - df_aod.config - enables Angle of Departure mode (this configuration requires antenna matrix)
>
> By default during build process df_aod.conf configuration is used.

This means that the settings in `df_aod.conf` should go directly into `prj.conf` and the config changes needed for `AoA` mode should go into a Kconfig fragment, for example `overlay-aoa.conf`.

That means you can remove:
```
if (NOT DF_MODE)
  message("DF_MODE is not being defined on the CMake command-line in the environment or by the app\n"
          "Setting DF_MODE to default value: AOD")
  set(DF_MODE "AOD")
endif()

if (DF_MODE STREQUAL "AOA")
  set(DF_CONF_FILE "${CMAKE_CURRENT_SOURCE_DIR}/df_aoa.conf")
elseif (DF_MODE STREQUAL "AOD")
  set(DF_CONF_FILE "${CMAKE_CURRENT_SOURCE_DIR}/df_aod.conf")
else()
  message(FATAL_ERROR "Selected unsupported DF_MODE: ${DF_MODE}.\n"
          "Please set DF_MODE parameter to one of following: AOA, AOD.")
endif()

if(NOT EXISTS ${DF_CONF_FILE})
  message(FATAL_ERROR "No such file or directory: DF_CONF_FILE: '${DF_CONF_FILE}'")
endif()
list(APPEND CONF_FILE ${DF_CONF_FILE})
list(APPEND CONF_FILE "prj.conf")
```

and then in the documentation simply write:
> To use Angle of Arrival mode, build this application as follows:
> `west build -b nrf52833dk_nrf52833  samples/bluetooth/df_per_adv_beacon -- -DOVERLAY_CONFIG=overlay-aoa.conf`

For more inspiration, you may take a look at echo server:
https://github.com/zephyrproject-rtos/zephyr/tree/master/samples/net/sockets/echo_server
https://docs.zephyrproject.org/latest/samples/net/sockets/echo_server/README.html

ppryga replied:
@tejlmand thank you! I didn't know about DOVERLAY_CONFIG. It looks like I was trying reinvent the wheel. I'll change it as suggested.

ppryga commented:
> call the sample `direction_beacon`, users can understand the sample better at first look.

The name is chosen on purpose, there will be another DF sample that will transmit CTE in connection mode.
I used the name to avoid having ambiguity in future.

ppryga-nordic committed:
Bluetooth: controller: Fix compilation errors related with DF code

Add missing declarations of functions that are compiled in when
CONFIG_BT_CTLR_ADV_EXT_PDU_EXTRA_DATA_MEMORY is selected
Add missing include of header file.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
Added commit that fixes build errors.
Implemented most of requested changes.
Waiting for response from @tejlmand.

mbolivar-nordic commented on subsys/bluetooth/controller/Kconfig.df:
```diff
@@ -120,4 +120,16 @@ config BT_CTLR_DF_MAX_ANT_SW_PATTERN_LEN
+	help
+	  Enable implicit initialization of GPIOs responsible for antenna
+	  switching during Radio initialization.
+	  If this feature is not enabled, Radio initiailization procedure will
```
```suggestion
	  If this feature is not enabled, the Radio initialization procedure will
```

ppryga replied:
corrected

mbolivar-nordic commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -26,46 +28,19 @@
 #define DF_PSEL_GPIO_NUM_MAX_IDX 7
 
-#if IS_ENABLED(CONFIG_BT_CTLR_DF_INIT_ANT_SEL_GPIOS)
-struct dfe_gpio_psel {
```
This PR adds this structure in the previous commit, only to remove it immediately. Can you please clean up the commits so this is easier to review and doesn't add and remove things like that? It looks like some of this commit should have been squashed into the previous commit instead during a rebase.

ppryga replied:
I've removed the code from previous commit.

mbolivar-nordic commented on dts/bindings/net/wireless/nordic,nrf-radio.yaml:
```diff
@@ -38,6 +38,17 @@ properties:
         Value zero means no antenna available. If provided value different than
         zero, it must be greater or equal to two.
 
+    dfe-pdu-ant:
```
I don't know what an antenna switch pattern is, but I wonder if the name of the property should be something like `dfe-ant-switch-pattern` or something like that?

tejlmand replied:
knowing that there also exists an `ANT+` wireless protocol, then I think using `ant` for antenna in a `net/wireless/nordic` file is bad naming.

Why is `antenna` so long that is has to be abbreviated ?

ppryga replied:
@tejlmand I'll change the name to `dfe-pdu-antenna`.

@mbolivar-nordic 
The `dfe-ant-switch-pattern` will not inform about actual purpose of the property. Direction Finding extension in Nordic Radio uses `switch-patterns` to drive hardware switches through GPIOs. That functionality is required to use multiple antennas when TX/RX Constant Tone Extension. 
There is max 40 `switch-patterns` stored in DFE. First of them is used by Radio peripheral (through DFE) to select antenna to TX/RX Bluetooth PDU (PDU that has attached CTE). BT spec. does not say anything about PDU antenna, so it is Nordic proprietary thing. Due to that I decided provide end user possibility to set PDU `switch-pattern` in DTS. It seems to me the best place to put it.

ppryga replied:
BTW I'll change `dfe-ant-num` to `dfe-antenna-num` to keep the same naming style.

mbolivar-nordic replied:
> The `dfe-ant-switch-pattern` will not inform about actual purpose of the property.

Thanks for confirming.

mbolivar-nordic commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -262,9 +228,26 @@ void radio_df_ant_switch_spacing_set_4us(void)
 }
 
-void radio_df_ant_switch_pattern_set(uint8_t pattern)
+void radio_df_ant_switch_pattern_set(uint8_t *patterns, uint8_t len)
```
The PS says "A minimum number of 3 patterns must be written." -- should this return an error if `len` is too short?

ppryga replied:
PS requires 3 patterns but pay attention that first pattern (`SWITCHPATTERN[0]`) is set with value provided by DTS. 
So there is a need for 2 more patterns and that is verified by HCI commands (according to BT spec. min. patterns number is 2). So there is no need to verify it again here. ULL handles that.

mbolivar-nordic replied:
>  that is verified by handling of HCI commands (according to BT spec. min. patterns number is 2). So there is no need to verify it again here. ULL handles that.

Great, thanks.

mbolivar-nordic reviewed (changes requested):
Hi, thanks for this. Looks great, just a few things.

tejlmand commented on samples/bluetooth/df_per_adv_beacon/README.rst:
```diff
@@ -0,0 +1,84 @@
+
+There are two build configuration provided with this sample:
+
+* df_aoa.config - enables Angle of Arrival mode
+* df_aod.config - enables Angle of Departure mode (this configuration requires
+  antenna matrix)
+
+By default during build process df_aod.conf configuration is used.
+
+To use Angle of Arrival mode, build this application as follows:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/bluetooth/df_cl_beacon
+   :host-os: unix
+   :board: nrf52833dk_nrf52833
+   :gen-args: DF_MODE=AOA
+   :goals: run
+   :compact:
```
Please update this to a single `prj.conf` and then a Kconfig fragment for `AoA` mode.

ppryga replied:
I've updated it as you have requested

tejlmand commented on samples/bluetooth/df_per_adv_beacon/boards/nrf52833dk_nrf52833.overlay:
```diff
@@ -0,0 +1,26 @@
+&radio {
+	status = "okay";
+	/* This is an example number of antennas that may be available
+	 * on atenna matrix board.
```
```suggestion
	 * on antenna matrix board.
```

ppryga replied:
Corrected

ppryga commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -23,12 +23,11 @@ static int init_reset(void);
 int lll_df_init(void)
 {
-	radio_df_ant_configure();
+	radio_df_ant_switching_gpios_cfg();
```
Nope, it is not. Thank you for spotting that!

ppryga-nordic committed:
dts: bindings: net: wireless: Change dfe-ant-num property name

Change dfe-ant-num propert name to dfe-antenna-num.
The change is conde to keep the same naming style
for all properies, like dfe-pdu-antenna property.

Signed-off-by: Piotr Pryga <piotr.pryga@nordicsemi.no>

ppryga commented:
All except one requested changes are done. Last one that requires additional PR: https://github.com/zephyrproject-rtos/zephyr/pull/32436 was implemented as proposed in comment: https://github.com/zephyrproject-rtos/zephyr/pull/32360#discussion_r580980395

mbolivar-nordic commented on subsys/bluetooth/controller/Kconfig.df:
```diff
@@ -92,7 +92,7 @@ config BT_CTLR_DF_CTE_RX
 	bool "Enable Connection CTE Response feature"
 	depends on BT_CTLR_DF_CTE_TX && BT_CONN
-	default y
+	default n
```
Style nit: we don't do an explicit `default n` since this is the default.

```suggestion
```

ppryga replied:
Thanks for spotting that, I've corrected it.

mbolivar-nordic reviewed (changes requested):
I just have one more style nit and then I am +1 to merge.

mbolivar-nordic commented:
CI errors look unrelated; maybe this needs a rebase?

ppryga commented:
Removed `default n` and rebased.

@mbolivar-nordic 
> CI errors look unrelated; maybe this needs a rebase?

There is some problem with qemu, it was reported but the issue is still open:
https://github.com/zephyrproject-rtos/zephyr/issues/32344

From time to time test fails...

mbolivar-nordic reviewed (approved):
LGTM. Thanks, nice work!

tejlmand commented on samples/bluetooth/df_per_adv_beacon/sample.yaml:
```diff
@@ -0,0 +1,7 @@
+  sample.bluetooth.df_beacon:
+    harness: bluetooth
+    platform_allow: nrf52833dk_nrf52833
+    tags: bluetooth
```
Could you add the `AoA` only mode to this, to make sure it's included when running twister:
```suggestion
    tags: bluetooth
  sample.bluetooth.df_beacon.aoa:
    extra_args: OVERLAY_CONFIG="overlay-aoa.conf" DTC_OVERLAY_FILE="boards/nrf52833dk_nrf52833_aoa_antenna.overlay"
    platform_allow: nrf52833dk_nrf52833
```

ppryga replied:
Added second configuration (`sample.bluetooth.df_beacon.aoa`) for twister build.

tejlmand reviewed (changes requested):
please include AoA only mode in sample.yaml

tejlmand commented on samples/bluetooth/df_per_adv_beacon/README.rst:
```diff
@@ -0,0 +1,79 @@
+To use Angle of Arrival mode only, build this application as follows:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/bluetooth/df_cl_beacon
```
Should be:
```suggestion
   :zephyr-app: samples/bluetooth/df_per_adv_beacon
```
if i'm not mistaken.

ppryga replied:
You are right, thank you for spotting that. I've already corrected it.

tejlmand reviewed (changes requested):
much better looking.

tejlmand commented:
> 
> 
> > call the sample `direction_beacon`, users can understand the sample better at first look.
> 
> @cvinayak
> The name is chosen on purpose, there will be another DF sample that will transmit CTE in connection mode.
> I used the name to avoid having ambiguity in future.

I'm aligned with @cvinayak here.
Good and clear names gives a better user experience, and doesn't beacon in the name indicates the advertising only nature ?

You could have:
- `direction_finding_beacon` thus implying advertising only.
- `direction_finding`, no beacon in the name, hinting this sample offer connection mode.

ppryga commented:
> >I'm aligned with @cvinayak here.
> Good and clear names gives a better user experience, and doesn't beacon in the name indicates the advertising only nature ?
>
> You could have:
>
>-  `direction_finding_beacon` thus implying advertising only. 
> - `direction_finding`, no beacon in the name, hinting this sample offer connection mode.

My plan is to create separate sample application for Direction Finding in connected mode. 
I'd like to keep samples simple and concentrated around one aspect of DF
- pair of application for periodic advertising DF (beacon, locator)
- pair of applications for DF in connected mode (beacon, locator)

To get clear distinction I've added information about periodic advertising in the name of the sample.
By use of `df_per_adv_beacon` I give clear information that DF is working with periodic advertising (will not work with legacy advertising).

At first attempt I'd used `df_cl_beacon` - Direction Finding Connectionless Beacon. After that I asked people if they could decrypt the name. Unfortunately they didn't. So I went for `df_per_adv_beacon` that one was more understandable.

This is my reasoning behind choice of the name.

tejlmand commented:
> My plan is to create separate sample application for Direction Finding in connected mode.
> I'd like to keep samples simple and concentrated around one aspect of DF

I still don't see the value in abbreviating direction finding to `df` in the folder name.
Paths are not that expensive, but makes a big difference to users browsing the folder structure.

Once inside `direction_finding_xyz` it's fine to abbreviate, as now the context is known to the user and `df` is more self explanatory.
So why not:
`direction_finding_per_adv_beacon` or `direction_per_adv_beacon`.

note, I won't block PR on this, but it seems i'm not the only one preferring a clearer name.

tejlmand reviewed (approved):
No more blocking change requests from me, but would like to hear opinion on the sample name from other reviewers.

Initially raised by @cvinayak here: https://github.com/zephyrproject-rtos/zephyr/pull/32360/#pullrequestreview-593922864

ppryga commented:
> I still don't see the value in abbreviating direction finding to df in the folder name.
Paths are not that expensive, but makes a big difference to users browsing the folder structure.

>Once inside direction_finding_xyz it's fine to abbreviate, as now the context is known to the user and df is more self explanatory.
So why not:
direction_finding_per_adv_beacon or direction_per_adv_beacon.

> >After that I asked people if they could decrypt the name.

> Already the fact that you have to ask people if they can decrypt the name is a clear sign that it's not well chosen.
There will be users out there with less context knowledge that have a harder time to know what this sample is about when browsing.

@tejlmand 
I agree with your opinion, it is not my private choice (I've already stated in one of former comments that my private preferences are long full names instead of abbreviations). Unfortunately there is line limit in github commit message (80 characters in line ?). If I name the sample `direction_finding_per_adv_beacon` then there is not much space to store some meaningful information in first line for future commits (if occur at all). If the limit is not enforced then I'd change the name to long one immediately.

tejlmand commented:
> Unfortunately there is line limit in github commit message (80 characters in line ?). If I name the sample direction_finding_per_adv_beacon then there is not much space to store some meaningful information in first line for future commits (if occur at all).

A github title should never restrict a good name.
You don't have to put the full path, just enough to make the commit title possible to identify there area.
`samples: direction finding: <text>` or `samples: bluetooth: df: <text>` could still be fine.

Just like this commit:
https://github.com/zephyrproject-rtos/zephyr/commit/93a8f5f10e4d9359b8f22071721199fb53d23a33
is just: `samples: boards: mec15xxevb: pm: <text>`
and not: `samples: boards: mec15xxevb_assy6853: power_management: <text>`
in the title.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_adv_internal.h:
```diff
@@ -10,6 +10,13 @@ int lll_adv_data_release(struct lll_adv_pdu *pdu);
 
 struct pdu_adv *lll_adv_pdu_alloc(struct lll_adv_pdu *pdu, uint8_t *idx);
 
+#if IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT_PDU_EXTRA_DATA_MEMORY)
```
Lets stop conditional compilation for declarations and only enclose definitions. Let the linker throw error when a symbol is undefined.

ppryga replied:
#if defined removed.

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -67,7 +67,10 @@ void lll_df_conf_cte_tx_enable(uint8_t type, uint8_t length,
 		radio_df_mode_set_aoa();
-	} else {
+	}
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) || \
+	IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX)
```
```suggestion
#if defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX) || \
	defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_RX)
```

ppryga replied:
corrected

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_df.c:
```diff
@@ -23,6 +23,7 @@
 #include "lll_adv.h"
 #include "ull_adv_types.h"
 #include "ull_adv_internal.h"
+#include "lll_adv_internal.h"
```
`lll_*_internal.h` should not be used in ULL files. I am working towards correcting it in https://github.com/zephyrproject-rtos/zephyr/pull/32613

See if you can split out vendor specifics in `lll_adv_internal.h`

ppryga replied:
The `#iclude lll_adv_internal.h` is required due to a single function; `lll_adv_sync_data_enqueue`. I can try to do a similar change as you did in https://github.com/zephyrproject-rtos/zephyr/pull/32613 but I do not want to change partially use of of `lll_xxx_internal.h` only partially. 

If I start to extract code that in your PR landed in `lll_adv_pdu.h` I'll be forced to change other files also. 
And it is not part of this commit goal.

In my humble opinion it is better to correct the use of header files in single PR (for all source files in controller).
If you strongly advise to do remove the `#iclude lll_adv_internal.h` from the `ull_df.c` I'll try to deliver it.

cvinayak replied:
> See if you can split out vendor specifics in lll_adv_internal.h

No, this was just a suggestion. Yes, can wait for a separate PR to address this, based on the outcome of #32613.

cvinayak commented on samples/bluetooth/df_per_adv_beacon/src/main.c:
```diff
@@ -0,0 +1,143 @@
+	.options = BT_LE_ADV_OPT_USE_TX_POWER,
+};
+
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX)
```
```suggestion
#if defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX)
```

ppryga replied:
corrected

cvinayak commented on samples/bluetooth/df_per_adv_beacon/src/main.c:
```diff
@@ -0,0 +1,143 @@
+struct bt_df_adv_cte_tx_param cte_params = {
+	.cte_len = CTE_LEN,
+	.cte_count = 1,
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX)
```
```suggestion
#if defined(CONFIG_BT_CTLR_DF_ANT_SWITCH_TX)
```

ppryga replied:
corrected

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -58,10 +66,14 @@ struct df_ant_cfg {
-BUILD_ASSERT((DT_PROP(RADIO, dfe_ant_num) >= DF_ANT_NUM_MIN), "Insufficient "
+BUILD_ASSERT((DT_PROP(RADIO, dfe_antenna_num) >= DF_ANT_NUM_MIN), "Insufficient "
 	     "number of antennas provided.");
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_INIT_ANT_SEL_GPIOS)
+BUILD_ASSERT((DT_PROP(RADIO, dfe_pdu_antenna) != DF_PDU_ANT_NOT_SET), "Missing "
+	     "antenna pattern used to select antenna for PDU Tx.");
+#endif /* CONFIG_BT_CTLR_DF_INIT_ANT_SEL_GPIOS */
```
```suggestion
BUILD_ASSERT(!IS_ENABLED(CONFIG_BT_CTLR_DF_INIT_ANT_SEL_GPIOS) || (DT_PROP(RADIO, dfe_pdu_antenna) != DF_PDU_ANT_NOT_SET), "Missing "
	     "antenna pattern used to select antenna for PDU Tx.");
```

ppryga replied:
corrected

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/hal/nrf5/radio/radio_df.c:
```diff
@@ -74,52 +86,77 @@ BUILD_ASSERT((DT_PROP(RADIO, dfe_ant_num) >= DF_ANT_NUM_MIN), "Insufficient "
-#endif /* CONFIG_BT_CTLR_DF_ANT_SWITCH_TX || CONFIG_BT_CTLR_DF_ANT_SWITCH_RX */
 }
 
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_INIT_ANT_SEL_GPIOS)
```
```suggestion
#if defined(CONFIG_BT_CTLR_DF_INIT_ANT_SEL_GPIOS)
```

ppryga replied:
corrected

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_df.c:
```diff
@@ -23,12 +23,12 @@ static int init_reset(void);
 {
-	radio_df_ant_configure();
-
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_INIT_ANT_SEL_GPIOS)
```
```suggestion
#if defined(CONFIG_BT_CTLR_DF_INIT_ANT_SEL_GPIOS)
```

ppryga replied:
corrected

cvinayak commented on subsys/bluetooth/host/direction.c:
```diff
@@ -190,6 +190,7 @@ static int hci_df_set_adv_cte_tx_enable(struct bt_le_ext_adv *adv,
 				   buf, NULL);
 }
 
+#if IS_ENABLED(CONFIG_BT_CTLR_DF_CONN_CTE_RSP)
```
```suggestion
#if defined(CONFIG_BT_CTLR_DF_CONN_CTE_RSP)
```

ppryga replied:
corrected

cvinayak commented:
> 
> 
> No more blocking change requests from me, but would like to hear opinion on the sample name from other reviewers.
> 
> Initially raised by @cvinayak here: #32360

Reading the Bluetooth Specification v5.2 Vol 1, Section 8, most commonly used term is "Direction Finding" and the methods being the Angle of Arrival (AoA) and Angle of Departure (AoD), and the transmitter (TX) and Receiver (RX) roles. There is no mention of `DF` in the Acronyms and Abbreviations. (I am against `DF` in the samples folder, which is the first match to `DFS` Dynamic Frequency Selection in the specification document).
 
Earlier proposed `direction_beacon` does not clearly relate to specification terminology.

Hence, my choices, going by the Direction Finding using Bluetooth Low Energy section in the specification document are:
`direction_finding_aoa_tx` or `direction_aoa_tx`
`direction_finding_aoa_rx` or `direction_aoa_rx`
`direction_finding_aod_tx` or `direction_aod_tx`
`direction_finding_aod_rx` or `direction_aod_rx`

For the connected roles, we can suffix `peripheral` or `central`.

ppryga commented:
Referring to sample name:

>A github title should never restrict a good name.
You don't have to put the full path, just enough to make the commit title possible to identify there area.
samples: direction finding: <text> or samples: bluetooth: df: <text> could still be fine.
@tejlmand Great to hear! In such situation I'll use long version of sample name.

@cvinayak 
If long name is possible (more it is welcome) then I will change it to full `direction_finding` (I'm very happy that it is possible!).

>Hence, my choices, going by the Direction Finding using Bluetooth Low Energy section in the specification document are:
`direction_finding_aoa_tx or direction_aoa_tx`
`direction_finding_aoa_rx or direction_aoa_rx`
`direction_finding_aod_tx or direction_aod_tx`
`direction_finding_aod_rx or direction_aod_rx`

Regarding to your proposals, I allow myself to not agree with your choices. Mentioned section does  not distinguish two methods the direction finding is realized according to protocol that are: connectionless (periodic advertising), connect. What more AoA and AoD terms are connected with transmitter and receiver roles in both possible realization ways. 
So it is possible to have e.g. `direction_finding_aoa_aod_tx` and `direction_finding_aoa_aod_rx`. 

> For the connected roles, we can suffix peripheral or central.
Also there is no realtation that peripheral role is responsible for e.g. direction finding transmission (see BT 5.2 Vol 6, Part D, section 6.16 Figure 6.39 "Master or slave-initiated Constant Tone Extension Request procedure – remote
device responds successfully").

I wish to have sample name that informs end user about feature it uses, realization mode (connectionless, connection) and role transmission, reception. So I suggest to use `direction_finding_connectionless_rx` for current `df_per_adv_beacon`. All parts of the new name have references in BT specification. For direction finding realized by connection I'll suggest to use `direction_finding_connection_rx`, `direction_finding_conection_tx`.

ppryga commented:
Implemented changes requested by @cvinayak  except: https://github.com/zephyrproject-rtos/zephyr/pull/32360#discussion_r582044506

Changed sample name to: `direction_finding_connectionless_tx`

tejlmand commented on samples/bluetooth/direction_finding_connectionless_tx/CMakeLists.txt:
```diff
@@ -0,0 +1,14 @@
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+
+project(df_beacon)
```
Maybe update this name now:
```suggestion
project(direction_finding_connectionless)
```

ppryga replied:
corrected

tejlmand commented on samples/bluetooth/direction_finding_connectionless_tx/sample.yaml:
```diff
@@ -0,0 +1,10 @@
+sample:
+  name: Direction Finding Connectionless Beacon
+tests:
+  sample.bluetooth.df_beacon:
```
maybe update this to reflect new name:
```suggestion
  sample.bluetooth.direction_finding_connectionless:
```

ppryga replied:
corrected

tejlmand commented on samples/bluetooth/direction_finding_connectionless_tx/sample.yaml:
```diff
@@ -0,0 +1,10 @@
+    harness: bluetooth
+    platform_allow: nrf52833dk_nrf52833
+    tags: bluetooth
+  sample.bluetooth.df_beacon.aoa:
```
maybe update this to reflect new name:
```suggestion
  sample.bluetooth.direction_finding_connectionless.aoa:
```

ppryga replied:
corrected

tejlmand reviewed (approved):
I propose to align to the new name a couple of places.
(PR still approved)

ppryga commented:
> I propose to align to the new name a couple of places.
(PR still approved)

I've changed all mentioned places. Thank you for checking it again!

carlescufi commented:
@ppryga please rebase

ppryga commented:
@carlescufi done :-)
